### ISSUE TITLE
Adapter-driven Architecture for WebBlocks

### ISSUE THREAD

**Issue by ebollens** [ucla-oarc-web/WebBlocks#16]
@bpollak raised a good issue during today's MWF User Group meeting that we should not require any particular UI framework in order to build WebBlocks. 

This can be broken into two requirements of the same overarching theme:
- It should be possible for the WebBlocks project to swap in a different baseline framework. We do not want to marry ourselves to any one particular framework, as over time we'll certainly see new frameworks crop up and others go unmaintained. This could happen to Bootstrap, just like any other - or for that matter, Bootstrap 3 could come out with significantly different markup style, just as they did when they released Bootstrap 2.
- It should be possible for a user of the framework to use a different underlying framework. If someone doesn't want to use Bootstrap but instead has their own framework that accomplishes the same things, they should be able to implement WebBlocks powered by their framework. This isn't to say that they wouldn't have to do any work, but rather that this work should be segregated in such a way that its clear what needs to be modified and what can stay as is.

From brainstorming with @ericlchiang in the office after the MWF User Group Meeting today, my first thought is some sort of an adapter layer.

**Comment by erutan** [comment]
I've never used Bootstrap - from what I've gleaned it is a lot more specific than most CSS frameworks out there, which makes sense for building a "branded" solution.  Switching around from a compass frameworks designed to be more agnostic/global shouldn't be too hard (or using it for core layout while keeping other elements of bootstrap).  I've used a liquid port of 960gs as well as zen grids on Drupal sites and it wouldn't be the end of the world to switch from one or another. 

re: adapter layer, having a structure.scss/layout.scss that was being @imported that contained classes containing framework mixins to be extended elsewhere seems like a solid approach to me.  You could run into some issues with how push/pull are handled between frameworks, but at least having it all one file makes it easier to manage. 

Some looks at cobbled together frameworks (though in the more hand waving CSS layout sense than hardcore framework sense) that are a mishmash of various solutions out there:

http://drupal.org/project/zen/
http://drupal.org/project/sasson

**Comment by ebollens** [comment]
@erutan, the focus of Bootstrap is mostly around actually providing specific UI elements, as opposed to many frameworks that stay focused mostly on layout, typography and other low-level issues. An example of the UI elements are things like buttons, menu bars, pill navs, forms, etc.

As for the adapter layer, I think what you're talking about is the subject of #5. What I'm referring to here is actually lower level architecturally, dealing with the ability to substitute one framework out in favor of another without modifying any WebBlocks provided files (except for configuration-type files).

Consider this definition in a file under the `lib` folder:

``` css
button {
    @extend .btn
}
```

Now, suppose we have a user that doesn't want to use Bootstrap. To modify this, they'd need to modify this file under `lib`. This means they have to modify a core framework file.

To get around this, envision if the call were something like:

```
button {
    @include entity-button;
}
```

Somewhere earlier in the stack, you'd first `@import "adapter/bootstrap"`, which would include a bunch of partials with mixins including one of the form:

```
@mixin entity-button {   
    @extend .btn;
}
```

You'd then be able to switch out which "adapter" is powering your entities simply by replacing `@import "adapter/bootstrap"` with something else like `@import "adapter/mycustom"`. You'd then be able to build your own custom adapters in a directory separate from any of the files defined by WebBlocks directly - meaning no merge conflicts and update issues.

I'm going to work up a commit that demonstrates this in basic.

**Comment by erutan** [comment]
My initial thought was to use a broader CSS framework and then write custom mixins for the low level stuff - but bootstrap sets you firmly in the direction of defining low level elements due it's inherent focus on structural markup (at the cost of making it responsive/compass integrated on your own). I'm obviously a little bit behind on the theorycrafting talks, sorry if I'm taking up time having you explain things out to me. ^^;

IMHO keeping the specific elements locked into blueprint/WebBlocks while allowing users to swap out the broader layout would seem to be a better fit than replacing the entire blueprint framework, given the focus on having a branded uniform appearance.  If someone wants to use a liquid blueprint/960/zen/susy whatever because they're used to that instead of bootstrap grids and there's a way to quick swap that'd make sense - but altering what typography, h/v rhythm etc is behind the framework and having it still uniformly appear seems like a mess and something better done behind the scenes when there is a new WebBlocks ready to be released.  Giving people classes and ids to use (and some mixins so they can pass through arguments if they want) that are set in stone between WebBlocks releases keeps visual branding simple, clean, and uniform - what grid system they want to use shouldn't affect that but might make them more comfortable.  Other layers of a framework very well could change things (and if they don't why would they be there?), and I don't think hacking them should be encouraged by individuals.

TL;DR - set it up so people can swap grid systems as they want with an @import in _base.scss and then overriding layout.scss, but everything is else is presented as documented black boxes that are simple to use, with flexibility given through guided mixin arguments on elements. For custom elements they can go part of the way there by @extending in .scss or adding in a class via html/css then building on top of that.

This said I know little about rake/ruby buildout architecture, just using SASS/Compass/design in general! Just my thoughts as a latecomer to the process. :)

**Comment by ebollens** [comment]
@erutan this issue came out of a discussion today during the MWF User Group call where @bpollak explained that UCSD has a different responsive framework they have that they might want to use as a base instead of Bootstrap. A good reason for this would be to cut down on excess definitions - if I've already defined a `.button` class in my framework, why package Bootstrap on top as well with `.btn` when I could just `@extend .button` instead.

The issue of swapping in and out grid systems is a different one. Most of the entities we're talking about are at a higher level than grids. For example, if I used the masthead markup (see `template/entity/header/masthead`), I don't even think about the grid. However, the grid is something that could definitely be tackled through an adapter too so that you could choose your gridding system by calling a mixin defined by whatever adapter you've loaded.

Also, I noticed that you're talking about giving people mixins to use. Ideally, WebBlocks is something that the end user sees as simple a `.css` file, not as a `.scss` file. This whole discussion of architecture is really about how we as framework developers (and advanced users looking to make serious tweaks like a central campus IT unit) work on the packages that are built. In most cases, an organization will tweak things for their organization, call `rake` and then just pass out the `.css` file instead.

And no worries on the explanations. It helps me think things out more thoroughly and provides documentation as we go as to why we made these decisions :)

**Comment by ebollens** [comment]
Commit 4e8b36d shows how this could work.

In the `src/config/_adapter.scss` file, you can specify which adapter you import. For example, you could import `adapter/bootstrap`. When you do this, the adapter is responsible for defining all the mixins required. In my example here, it defines the mixin `-entity-button`. I'm using the hyphen as a prefix to denote that it's an adapter-related mixin. Then, all of the libraries under `lib/entity/button` use this mixin to define the button's framework-powered properties (such as `@extend .btn`).

I'm unconvinced whether this is the "right" way to do it, but it's one possible route. The key here is to decouple the `lib`, which defines the entities in the framework, from the adapter, which specifies the specific mapping required to achieve this given a specific framework. Definitely more complex than the scheme I presented formerly in #5, but it solves the framework problem.

**Comment by erutan** [comment]
It's an interesting shift in perspective for me - I've grown used to being a technical team of one (with the pros and cons that come with that) whether it was at UCSF or doing client work now. :)  Working on an existing site where I can't use SCSS annoys me at this point! 

Tying the framework with centralized hosting helps keep the cost down on custom implementations down and enforce consistency upon them regardless of who implemented them, which would increase the interest in the system as well as making future changes easier (rebuild when your framework shifts, boom, you're most of the way there to a new visual identity).

Having a different UC use their own framework behind this one makes sense. I was thinking of it from a single campus perspective, where the main advantage of abstraction are future upgrades on the central/architectural level.

**Comment by ebollens** [comment]
Ah, I gotcha. Yeah, the goal of this toolkit is to allow any institution looking to deploy it to be able to do so sustainably, as well as for our core dev group to maintain it sustainably. Further, MWF has shown the benefits of a federated architecture (central CSS, JS and compressor tools hosted for a distributed set of apps), hence why we're considering the use case of people that just want to consume the CSS as well as those that might want to modify the SCSS or even use a completely different framework.

**Comment by ebollens** [comment]
I am closing this, as we seem to be in agreement on this approach.
